### ISSUE TITLE
pylint reports false 'import-error' since version 3.3.2 in conjunction with namespace packages

### ISSUE THREAD

**Issue by b-morgenthaler** [pylint-dev/pylint#10185]
### Bug description

```python
Pylint reports (since version 3.3.2) a false `import-error` when used in conjunction with Python's "src" layout and namespace packages. Please see attached minimal repository to reproduce the error. If the same repo with the same config is linted with version 3.3.1, no false `import-error` occurs.
```

### Configuration

```ini
[tool.pylint.main]
# Return non-zero exit code if any of these messages/categories are detected,
# even if score is above --fail-under value. Syntax same as enable. Messages
# specified are enabled, while categories only check already-enabled messages.
fail-on = ["F", "E", "W", "C", "R"]

# Specify a score threshold under which the program will exit with error.
fail-under = 10

# Add files or directories matching the regular expressions patterns to the
# ignore-list. The regex matches against paths and can be in Posix or Windows
# format. Because '\\' represents the directory delimiter on Windows systems, it
# can't be used as an escape character.
# ignore-paths =

# Files or directories matching the regular expression patterns are skipped. The
# regex matches against base names, not paths. The default value ignores Emacs
# file locks
ignore-patterns = ["^\\.#"]

# List of module names for which member attributes should not be checked and will
# not be imported (useful for modules/projects where namespaces are manipulated
# during runtime and thus existing member attributes cannot be deduced by static
# analysis). It supports qualified module names, as well as Unix pattern
# matching.
# ignored-modules =

# Minimum Python version to use for version dependent checks. Will default to the
# version used to run pylint.
py-version = "3.10"

# Add paths to the list of the source roots. Supports globbing patterns. The
# source root is an absolute path or a path relative to the current working
# directory used to determine a package namespace for modules located under the
# source root.
source-roots = "src"

[tool.pylint.basic]
# Naming style matching correct argument names.
argument-naming-style = "snake_case"

# Regular expression matching correct argument names. Overrides argument-naming-
# style. If left empty, argument names will be checked with the set naming style.
# argument-rgx =

# Naming style matching correct attribute names.
attr-naming-style = "snake_case"

# Regular expression matching correct attribute names. Overrides attr-naming-
# style. If left empty, attribute names will be checked with the set naming
# style.
# attr-rgx =

# Bad variable names which should always be refused, separated by a comma.
bad-names = ["foo", "bar", "baz", "toto", "tutu", "tata"]

# Bad variable names regexes, separated by a comma. If names match any regex,
# they will always be refused
# bad-names-rgxs =

# Naming style matching correct class attribute names.
class-attribute-naming-style = "any"

# Regular expression matching correct class attribute names. Overrides class-
# attribute-naming-style. If left empty, class attribute names will be checked
# with the set naming style.
# class-attribute-rgx =

# Naming style matching correct class constant names.
class-const-naming-style = "UPPER_CASE"

# Regular expression matching correct class constant names. Overrides class-
# const-naming-style. If left empty, class constant names will be checked with
# the set naming style.
# class-const-rgx =

# Naming style matching correct class names.
class-naming-style = "PascalCase"

# Regular expression matching correct class names. Overrides class-naming-style.
# If left empty, class names will be checked with the set naming style.
# class-rgx =

# Naming style matching correct constant names.
const-naming-style = "UPPER_CASE"

# Regular expression matching correct constant names. Overrides const-naming-
# style. If left empty, constant names will be checked with the set naming style.
# const-rgx =

# Minimum line length for functions/classes that require docstrings, shorter ones
# are exempt.
docstring-min-length = -1

# Naming style matching correct function names.
function-naming-style = "snake_case"

# Regular expression matching correct function names. Overrides function-naming-
# style. If left empty, function names will be checked with the set naming style.
# function-rgx =

# Good variable names which should always be accepted, separated by a comma.
good-names = ["i", "j", "k", "ex", "Run", "_"]

# Good variable names regexes, separated by a comma. If names match any regex,
# they will always be accepted
# good-names-rgxs =

# Include a hint for the correct naming format with invalid-name.
# include-naming-hint =

# Naming style matching correct inline iteration names.
inlinevar-naming-style = "any"

# Regular expression matching correct inline iteration names. Overrides
# inlinevar-naming-style. If left empty, inline iteration names will be checked
# with the set naming style.
# inlinevar-rgx =

# Naming style matching correct method names.
method-naming-style = "snake_case"

# Regular expression matching correct method names. Overrides method-naming-
# style. If left empty, method names will be checked with the set naming style.
# method-rgx =

# Naming style matching correct module names.
module-naming-style = "snake_case"

# Regular expression matching correct module names. Overrides module-naming-
# style. If left empty, module names will be checked with the set naming style.
# module-rgx =

# Colon-delimited sets of names that determine each other's naming style when the
# name regexes allow several styles.
# name-group =

# Regular expression which should only match function or class names that do not
# require a docstring.
no-docstring-rgx = "^_"

# List of decorators that produce properties, such as abc.abstractproperty. Add
# to this list to register other decorators that produce valid properties. These
# decorators are taken in consideration only for invalid-name.
property-classes = ["abc.abstractproperty"]

# Regular expression matching correct type alias names. If left empty, type alias
# names will be checked with the set naming style.
# typealias-rgx =

# Regular expression matching correct type variable names. If left empty, type
# variable names will be checked with the set naming style.
# typevar-rgx =

# Naming style matching correct variable names.
variable-naming-style = "snake_case"

# Regular expression matching correct variable names. Overrides variable-naming-
# style. If left empty, variable names will be checked with the set naming style.
# variable-rgx =

[tool.pylint.classes]
# Warn about protected attribute access inside special methods
# check-protected-access-in-special-methods =

# List of method names used to declare (i.e. assign) instance attributes.
defining-attr-methods = ["__init__", "__new__", "setUp", "asyncSetUp", "__post_init__"]

# List of member names, which should be excluded from the protected access
# warning.
exclude-protected = ["_asdict", "_fields", "_replace", "_source", "_make", "os._exit"]

# List of valid names for the first argument in a class method.
valid-classmethod-first-arg = ["cls"]

# List of valid names for the first argument in a metaclass class method.
valid-metaclass-classmethod-first-arg = ["mcs"]

[tool.pylint.design]
# List of regular expressions of class ancestor names to ignore when counting
# public methods (see R0903)
# exclude-too-few-public-methods =

# List of qualified class names to ignore when counting class parents (see R0901)
# ignored-parents =

# Maximum number of arguments for function / method.
max-args = 6

# Maximum number of positional arguments for function / method (see R0917).
max-positional-arguments = 6

# Maximum number of attributes for a class (see R0902).
max-attributes = 8

# Maximum number of boolean expressions in an if statement (see R0916).
max-bool-expr = 5

# Maximum number of branch for function / method body.
max-branches = 12

# Maximum number of locals for function / method body.
max-locals = 15

# Maximum number of parents for a class (see R0901).
max-parents = 7

# Maximum number of public methods for a class (see R0904).
max-public-methods = 20

# Maximum number of return / yield for function / method body.
max-returns = 6

# Maximum number of statements in function / method body.
max-statements = 50

# Minimum number of public methods for a class (see R0903).
min-public-methods = 2

[tool.pylint.exceptions]
# Exceptions that will emit a warning when caught.
overgeneral-exceptions = ["builtins.BaseException", "builtins.Exception"]

[tool.pylint.format]
# Expected format of line ending, e.g. empty (any line ending), LF or CRLF.
# expected-line-ending-format =

# Regexp for a line that is allowed to be longer than the limit.
ignore-long-lines = "^\\s*(# )?<?https?://\\S+>?$"

# Number of spaces of indent required inside a hanging or continued line.
indent-after-paren = 4

# String used as indentation unit. This is usually "    " (4 spaces) or "\t" (1
# tab).
indent-string = "    "

# Maximum number of characters on a single line.
max-line-length = 100  # shall be greater than black's config

# Maximum number of lines in a module.
max-module-lines = 1000

# Allow the body of a class to be on the same line as the declaration if body
# contains single statement.
# single-line-class-stmt =

# Allow the body of an if to be on the same line as the test if there is no else.
# single-line-if-stmt =

[tool.pylint.imports]
# List of modules that can be imported at any level, not just the top level one.
# allow-any-import-level =

# Allow explicit reexports by alias from a package __init__.
# allow-reexport-from-package =

# Allow wildcard imports from modules that define __all__.
# allow-wildcard-with-all =

# Deprecated modules which should not be used, separated by a comma.
# deprecated-modules =

# Output a graph (.gv or any supported image format) of external dependencies to
# the given file (report RP0402 must not be disabled).
# ext-import-graph =

# Output a graph (.gv or any supported image format) of all (i.e. internal and
# external) dependencies to the given file (report RP0402 must not be disabled).
# import-graph =

# Output a graph (.gv or any supported image format) of internal dependencies to
# the given file (report RP0402 must not be disabled).
# int-import-graph =

# Force import order to recognize a module as part of the standard compatibility
# libraries.
# known-standard-library =

# Force import order to recognize a module as part of a third party library.
known-third-party = ["enchant"]

# Couples of modules and preferred modules, separated by a comma.
# preferred-modules =

[tool.pylint.logging]
# The type of string formatting that logging methods do. `old` means using %
# formatting, `new` is for `{}` formatting.
logging-format-style = "old"

# Logging modules to check that the string format arguments are in logging
# function parameter format.
logging-modules = ["logging"]

[tool.pylint."messages control"]
# Only show warnings with the listed confidence levels. Leave empty to show all.
# Valid levels: HIGH, CONTROL_FLOW, INFERENCE, INFERENCE_FAILURE, UNDEFINED.
confidence = ["HIGH", "CONTROL_FLOW", "INFERENCE", "INFERENCE_FAILURE", "UNDEFINED"]

# Disable the message, report, category or checker with the given id(s). You can
# either give multiple identifiers separated by comma (,) or put this option
# multiple times (only on the command line, not in the configuration file where
# it should appear only once). You can also use "--disable=all" to disable
# everything first and then re-enable specific checks. For example, if you want
# to run only the similarities checker, you can use "--disable=all
# --enable=similarities". If you want to run only the classes checker, but have
# no Warning level messages displayed, use "--disable=all --enable=classes
# --disable=W".
disable = ["missing-module-docstring"]

# Enable the message, report, category or checker with the given id(s). You can
# either give multiple identifier separated by comma (,) or put this option
# multiple time (only on the command line, not in the configuration file where it
# should appear only once). See also the "--disable" option for examples.
# enable =

[tool.pylint.method_args]
# List of qualified names (i.e., library.method) which require a timeout
# parameter e.g. 'requests.api.get,requests.api.post'
timeout-methods = ["requests.api.delete", "requests.api.get", "requests.api.head", "requests.api.options", "requests.api.patch", "requests.api.post", "requests.api.put", "requests.api.request"]

[tool.pylint.miscellaneous]
# List of note tags to take in consideration, separated by a comma.
notes = ["FIXME", "XXX", "TODO"]

# Regular expression of note tags to take in consideration.
# notes-rgx =

[tool.pylint.refactoring]
# Maximum number of nested blocks for function / method body
max-nested-blocks = 5

# Complete name of functions that never returns. When checking for inconsistent-
# return-statements if a never returning function is called then it will be
# considered as an explicit return statement and no message will be printed.
never-returning-functions = ["sys.exit", "argparse.parse_error"]

# Let 'consider-using-join' be raised when the separator to join on would be non-
# empty (resulting in expected fixes of the type: ``"- " + " - ".join(items)``)
suggest-join-with-non-empty-separator = true

[tool.pylint.reports]
# Python expression which should return a score less than or equal to 10. You
# have access to the variables 'fatal', 'error', 'warning', 'refactor',
# 'convention', and 'info' which contain the number of messages in each category,
# as well as 'statement' which is the total number of statements analyzed. This
# score is used by the global evaluation report (RP0004).
evaluation = "max(0, 0 if fatal else 10.0 - ((float(5 * error + warning + refactor + convention) / statement) * 10))"

# Set the output format. Available formats are: text, parseable, colorized, json2
# (improved json format), json (old json format) and msvs (visual studio). You
# can also give a reporter class, e.g. mypackage.mymodule.MyReporterClass.
output-format = "colorized,parseable:pylint.txt"

# Tells whether to display a full report or only the messages.
reports = false

# Activate the evaluation score.
score = true

[tool.pylint.similarities]
# Comments are removed from the similarity computation
ignore-comments = true

# Docstrings are removed from the similarity computation
ignore-docstrings = true

# Imports are removed from the similarity computation
ignore-imports = true

# Signatures are removed from the similarity computation
ignore-signatures = true

# Minimum lines number of a similarity.
min-similarity-lines = 4

[tool.pylint.spelling]
# Limits count of emitted suggestions for spelling mistakes.
max-spelling-suggestions = 4

# Spelling dictionary name. No available dictionaries : You need to install both
# the python package and the system dependency for enchant to work.
# spelling-dict =

# List of comma separated words that should be considered directives if they
# appear at the beginning of a comment and should not be checked.
spelling-ignore-comment-directives = "fmt: on,fmt: off,noqa:,noqa,nosec,isort:skip,mypy:"

# List of comma separated words that should not be checked.
# spelling-ignore-words =

# A path to a file that contains the private dictionary; one word per line.
# spelling-private-dict-file =

# Tells whether to store unknown words to the private dictionary (see the
# --spelling-private-dict-file option) instead of raising a message.
# spelling-store-unknown-words =

[tool.pylint.typecheck]
# List of decorators that produce context managers, such as
# contextlib.contextmanager. Add to this list to register other decorators that
# produce valid context managers.
contextmanager-decorators = ["contextlib.contextmanager"]

# List of members which are set dynamically and missed by pylint inference
# system, and so shouldn't trigger E1101 when accessed. Python regular
# expressions are accepted.
# generated-members =

# Tells whether missing members accessed in mixin class should be ignored. A
# class is considered mixin if its name matches the mixin-class-rgx option.
# Tells whether to warn about missing members when the owner of the attribute is
# inferred to be None.
ignore-none = true

# This flag controls whether pylint should warn about no-member and similar
# checks whenever an opaque object is returned when inferring. The inference can
# return multiple potential results while evaluating a Python object, but some
# branches might not be evaluated, which results in partial inference. In that
# case, it might be useful to still emit no-member and other checks for the rest
# of the inferred objects.
ignore-on-opaque-inference = true

# List of symbolic message names to ignore for Mixin members.
ignored-checks-for-mixins = ["no-member", "not-async-context-manager", "not-context-manager", "attribute-defined-outside-init"]

# List of class names for which member attributes should not be checked (useful
# for classes with dynamically set attributes). This supports the use of
# qualified names.
ignored-classes = ["optparse.Values", "thread._local", "_thread._local", "argparse.Namespace"]

# Show a hint with possible names when a member name was not found. The aspect of
# finding the hint is based on edit distance.
missing-member-hint = true

# The minimum edit distance a name should have in order to be considered a
# similar match for a missing member name.
missing-member-hint-distance = 1

# The total number of similar names that should be taken in consideration when
# showing a hint for a missing member.
missing-member-max-choices = 1

# Regex pattern to define which classes are considered mixins.
mixin-class-rgx = ".*[Mm]ixin"

# List of decorators that change the signature of a decorated function.
# signature-mutators =

[tool.pylint.variables]
# List of additional names supposed to be defined in builtins. Remember that you
# should avoid defining new builtins when possible.
# additional-builtins =

# Tells whether unused global variables should be treated as a violation.
allow-global-unused-variables = true

# List of names allowed to shadow builtins
# allowed-redefined-builtins =

# List of strings which can identify a callback function by name. A callback name
# must start or end with one of those strings.
callbacks = ["cb_", "_cb"]

# A regular expression matching the name of dummy variables (i.e. expected to not
dummy-variables-rgx = "_+$|(_[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)|dummy|^ignored_|^unused_"

# Argument names that match this expression will be ignored.
ignored-argument-names = "_.*|^ignored_|^unused_"

# Tells whether we should check for unused import in __init__ files.
# init-import =

# List of qualified module names which can have objects that can redefine
# builtins.
redefining-builtins-modules = ["six.moves", "past.builtins", "future.builtins", "builtins", "io"]
```

### Command used

```shell
pylint src.namespace.library
```

### Pylint output

```python
(pylint_test) PS D:\git\minimal_pylint_example> pylint src.namespace.library
************* Module src.namespace.library
src\namespace\library\__init__.py:1:0: E0401: Unable to import 'src.namespace.library.library' (import-error)
```

### Expected behavior

No `import-error` 

[minimal_pylint_example.zip](https://github.com/user-attachments/files/18413190/minimal_pylint_example.zip)

### Pylint version

```shell
pylint 3.3.2 and pylint 3.3.3
astroid 3.3.8
python 3.10.15
```

### OS / Environment

Windows 10

### Additional dependencies

```python

```

**Comment by echoix** [comment]
I think there's a usage problem with either your command, or your config file.

If you set source root to "src", that would mean that you'd want the namespace to start inside that folder. Just like if you'd cd into src and see namespace.library as importable.

If you want to keep your command called as is, and your imports really look like "import src.namespace.library", then in your config it should not have a source root configured like that.


For you, it is a regression as you adapted your Pylint call for the non-working source root config option

**Comment by o-moe** [comment]
Thanks, @echoix for this explanation.

I explicitly want the source repo to use the "src" layout according to https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/. To my understanding this is what may minimal example shows. That's the reason why I configured pylint with `source-roots = "src"`.

Inside this source root, there's a namespace package according to PEP420 (without `__init__.py`) so that my library can be imported by `namespace.library` by apps. Which is perfectly working within my own private projects.

Calling pylint with `src.namespace.library` up to the point of pylint 3.3.2 were the only call that worked for me given the layout. Is this what you mean with "it is a regression for me"?

Should have been the correct way of calling pylint with this configuration `pylint namespace.library` all along and this was a bug that I "worked around"?

If yes, sorry for the confusion and I will adapt my linter script and pylint dependency.

**Comment by echoix** [comment]
> Calling pylint with src.namespace.library up to the point of pylint 3.3.2 were the only call that worked for me given the layout. Is this what you mean with "it is a regression for me"?
> 
> Should have been the correct way of calling pylint with this configuration pylint namespace.library all along and this was a bug that I "worked around"?

Yes, that's what I think so. It was impossible to use correctly before, especially if there was more than one level deep, or multiple source roots. (With one of the source roots being used as an installed package in the second one). Pylint had to be called with different working directories and config, and couldn't just know the correct module name for imports.


You can call Pylint with files or modules. For modules, you'd want to call it with the same module name as when users import your library in their files.

**Comment by Julfried** [comment]
Thanks for the clarification, @echoix. Your explanation is spot-on and aligns with my original intent behind the fix in #10036.

Before this change, pylint didn’t handle source-roots correctly, especially in cases involving the src layout. Users had to adapt by calling pylint with non-standard paths (e.g., src.namespace.library), which was more of a workaround than correct usage. The goal of my PR was to fix this behavior so that configurations like source-roots = "src" work as expected, allowing users to call pylint namespace.library as it would be imported in actual code.

I opened #10173 after comments from @webknjaz, as they described similar challenges with source root handling and implicit namespaces. If this issue proves that the behavior works as described for PEP 420 namespaces, it might also help resolve #10173.

@o-moe, could you clarify whether the issue here behaves correctly with implicit namespaces as specified in PEP 420?

Let me know if further clarification is needed—I’m happy to assist!

**Comment by o-moe** [comment]
@echoix If your statement regarding how to call modules with pylint is the desired/expected behavior

> You can call Pylint with files or modules. For modules, you'd want to call it with the same module name as when users import your library in their files.

Then I can confirm that it was broke before v3.3.2 (without me actually knowing it) and is working since then. Sorry for the confusion/wrong issue.

@Julfried To my understanding, the example I provide, uses Python's src layout and has a namespace "namespace" according to PEP420 and from v3.3.2 on,  pylint is able to call the library module within the namespace by `pylint namespace.library` which seems to be the desired behavior. To be precise: the former (wrong) way of calling the module by `src.namespace.library` does not work anymore - so it's not backwards compatible (even though this call seems to be wrong, it's worth pointing out imho).

PEP420 also allows for nested namespaces which I did not use in my example and namespaces should be independent from the layout used (src and flat). Do you want me to verify these scenarios as well?

**Comment by Julfried** [comment]
Thank you for the clarification, @o-moe! I’m glad to hear the fix aligns with PEP 420 and works as intended from v3.3.2 onward. While the old behavior is no longer supported, it’s worth noting it was technically incorrect. The current implementation aligns with PEP 420, which is a step in the right direction. If you’re able to verify other scenarios like nested namespaces (e.g., namespace.subnamespace.library) or usage in flat vs. src layouts, it would be very helpful for confirming full PEP 420 compliance.

**Comment by o-moe** [comment]
I extended my minimal example to include nested namespaces according to PEP420 and used `src` as well as `flat` layout. Both, version 3.3.2 and the current HEAD of main worked like the expectation given by @echoix before.

Worth noting:

- in a flat project layout, it would not make a difference if `source-roots` were defined (even if not existing). There were no complaints/warnings/errors - it just worked.
- in a src project layout, it made a difference: without defining the `source-roots` properly, a `no module named` error occurred
- the "old/wrong" behavior works indeed with version 3.3.2 (and main HEAD) and my example, when the import of the `library.Library` class from `src/namespace/library/__init__.py` is removed.

For reference, here's a tree of my minimal example archive attached above:

```shell
(.venv) bm@mba minimal_pylint_example % tree
.
├── LICENSE
├── README.md
├── pylint.txt
├── pyproject.toml
└── src
    └── namespace
        └── library
            ├── __init__.py
            ├── copyright.py
            ├── library.py
            └── py.typed

4 directories, 8 files

```

**Comment by Julfried** [comment]
Thanks for verifying, @o-moe! Since the behavior now aligns with PEP 420 and works as expected from v3.3.2 onward, I believe this issue can be closed. The previous behavior (`pylint src.namespace.library`) was incorrect, and the fix ensures that `source-roots` work properly. #10147 and #10173 are probably related and can also be closed

@Pierre-Sassoulas, let me know if you think further clarification is needed, this looks to be resolved.

Thanks again for the detailed testing, @o-moe!

**Comment by Pierre-Sassoulas** [comment]
Thank you for investigating everyone !
